tools/codegen/api/unboxing.py

```diff
@@ -0,0 +1,192 @@
+#           } else {
+#               dtype_opt_out = c10::optional<at::ScalarType>();
+#           }
+#       ```
+#
+# - Unboxed Kernel Call
+#   ```cpp
+#       auto result_ = torch::empty(
+#           size_list_out,
+#           names_opt_out,
+#           options,
+#           memory_format_opt_out
+#       );
+#   ```
+#
+# - Push Result Back to Stack
+#   ```cpp
+#       drop(stack, 7);
+#       pack(stack, std::move(result_));
+#   ```
+connector = "\n\t"
+
+
+# Convert all the arguments in a NativeFunction to C++ code, including TensorOptions.
+def convert_arguments(args: List[Argument]) -> Tuple[List[Expr], List[str]]:
+    argument_str = "c10::IValue {arg_name} = std::move(peek(stack, {pos}, {args_num}));"
+    expr_list = []
+    pop_ivalue = []
+    code_list = []
+    for i, arg in enumerate(args):
+        pop_ivalue.append(argument_str.format(arg_name=arg.name, pos=i, args_num=len(args)))
+        expr, code = argumenttype_ivalue_convert(arg.type, arg.name, mutable=arg.is_write)
+        expr_list.append(expr)
+        code_list.extend(code)
+    pop_ivalue.append("")
+    return expr_list, pop_ivalue + code_list
+
+
+# Take an argument in JIT type format, returns the C++ code to convert an ivalue from stack to corresponding C++ type.
+def argumenttype_ivalue_convert(t: Type, arg_name: str, *, mutable: bool = False) -> Tuple[Expr, List[str]]:
+    nctype = argumenttype_type(t=t, mutable=mutable, binds=arg_name)
+    ctype = nctype.cpp_type(strip_ref=True)
+
+    if isinstance(t, BaseType):
+        out_name = f"{arg_name}_base"
+        code = [f"{ctype} {arg_name}_base = {arg_name}.to<{ctype}>();"]
+    elif isinstance(t, OptionalType):
+        out_name = arg_name + "_opt_out"
+        code = _gen_code_optional_type(arg_name=arg_name, out_name=out_name, t=t, ctype=ctype)
+    elif isinstance(t, ListType):
+        out_name = arg_name + "_list_out"
+        code = _gen_code_list_type(arg_name=arg_name, out_name=out_name, t=t, ctype=ctype)
+    else:
+        raise Exception(f"Cannot handle type {t}. arg_name: {arg_name}")
+    expr = Expr(expr=out_name, type=nctype)
+    return expr, code
+
+
+def _gen_code_optional_type(arg_name: str, out_name: str, t: OptionalType, ctype: str) -> List[str]:
+    in_name = arg_name + "_opt_in"
+    res_expr, res_code = argumenttype_ivalue_convert(t.elem, in_name)
+    return f"""
+c10::optional<c10::IValue> {arg_name + "_opt"} = {arg_name}.toOptional<c10::IValue>();
+{ctype} {out_name};
+if ({arg_name + "_opt"}.has_value()) {{
+    const c10::IValue {in_name} = {arg_name + "_opt"}.value();
+    {connector.join(res_code)}
+    {out_name} = {ctype}({res_expr.expr});
+}} else {{
+    {out_name} = {ctype}();
+}}
+        """.split("\n")
+
+
+def _gen_code_list_type(arg_name: str, out_name: str, t: ListType, ctype: str) -> List[str]:
+    in_name = arg_name + "_list_in"
+    elem_name = arg_name + "_elem"
+    code = [f"const c10::List<c10::IValue> {in_name} = {arg_name}.toList();"]
```

Comment:
Hmm... I think this is probably okay

We generally want to avoid hardcoding C++ types as strings into random places in the codegen, so there's one source of truth for everything. There's already a notion of "The C++ type for list" in `tools/codegen/api/types.py`. But that file doesn't know what `IValue`'s are, and I'm not sure that we should add that yet. No action necessary though in this PR.